## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-zoned.sh | :x: [more...](#e4b11389-c0e1-43c2-85b3-cf78e1f04899) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#078388a4-a63a-4cd0-ba01-20f8d924c975) |
<a name="078388a4-a63a-4cd0-ba01-20f8d924c975"></a>
### 12161-64b4d7d/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:32199
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 881a46c6d446:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... FAILED in 31 seconds
{
  "name": "awscli",
  "duration": 2709,
  "function": "aws --endpoint-url http://minio-dev8.minio.io:32199 s3api copy-object --bucket awscli-mint-test-bucket-25014 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-25014/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 50 seconds
(7/15) Running minio-dotnet tests ... done in 34 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 28 seconds
{
  "args": {},
  "duration": 606,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 2 minutes and 8 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 120,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-kopi72, objectName = minio-java-test-2fk691f-copy, resource = /minio-java-test-kopi72/minio-java-test-2fk691f-copy, requestId = 1679831B93C7D908, hostId = 2377bbf9-8dbe-4fcd-a420-d8ffd21bf628)\nrequest={method=PUT, url=http://minio-dev8.minio.io:32199/minio-java-test-kopi72/minio-java-test-2fk691f-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-1oubea2/minio-java-test-2fk691f\nHost: minio-dev8.minio.io:32199\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210426T204135Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210426/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 416\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Mon, 26 Apr 2021 20:41:35 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 1679831B93C7D908\nX-Xss-Protection: 1; mode=block\nDate: Mon, 26 Apr 2021 20:41:35 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 46 seconds
(9/15) Running minio-py tests ... done in 2 minutes and 44 seconds
(10/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "2904",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-2994/s3cmd-test-object-1447 75ce976f-a105-46b5-9fe7-00ad7b70c3d1\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 75ce976f-a105-46b5-9fe7-00ad7b70c3d1 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 6 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
<a name="e4b11389-c0e1-43c2-85b3-cf78e1f04899"></a>
### 12161-64b4d7d/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31612
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1a57def8593e:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... done in 2 minutes and 11 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 1 minutes and 3 seconds
(8/15) Running minio-dotnet tests ... done in 49 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 9 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 51 seconds
(11/15) Running minio-js tests ... done in 45 seconds
(12/15) Running minio-py tests ... FAILED in 2 minutes and 44 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-ba67acc0-8a2c-48bc-9e8b-4749d04c5816",
    "object_name": "15552bff-2183-4723-ac9f-877432b576d5"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-ba67acc0-8a2c-48bc-9e8b-4749d04c5816, request_id: 16798360DD7097E3, host_id: f7701a03-6bd9-426b-a92d-f7a1519af9e8, bucket_name: minio-py-test-ba67acc0-8a2c-48bc-9e8b-4749d04c5816",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1640, in test_thread_safe\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-ba67acc0-8a2c-48bc-9e8b-4749d04c5816, request_id: 16798360DD7097E3, host_id: f7701a03-6bd9-426b-a92d-f7a1519af9e8, bucket_name: minio-py-test-ba67acc0-8a2c-48bc-9e8b-4749d04c5816\n",
  "duration": 6263
}
(12/15) Running s3cmd tests ... done in 17 seconds
(13/15) Running s3select tests ... done in 6 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/erasure-healing.go
+++ b/cmd/erasure-healing.go
@@ -505,6 +505,11 @@ func (er erasureObjects) healObject(ctx context.Context, bucket string, object s
 		// record the index of the updated disks
 		partsMetadata[i].Erasure.Index = i + 1
 
+		// dataDir should be empty when
+		// - transitionStatus is complete and not in restored state
+		if partsMetadata[i].TransitionStatus == lifecycle.TransitionComplete && !isRestoredObjectOnDisk(partsMetadata[i].Metadata) {
+			partsMetadata[i].DataDir = ""
+		}
 		// Attempt a rename now from healed data to final location.
 		if err = disk.RenameData(ctx, minioMetaTmpBucket, tmpID, partsMetadata[i], bucket, object); err != nil {
 			logger.LogIf(ctx, err)

--- a/cmd/erasure-multipart.go
+++ b/cmd/erasure-multipart.go
@@ -575,7 +575,7 @@ func (er erasureObjects) PutObjectPart(ctx context.Context, bucket, object, uplo
 		PartNumber:   partID,
 		ETag:         md5hex,
 		LastModified: fi.ModTime,
-		Size:         fi.Size,
+		Size:         n,
 		ActualSize:   data.ActualSize(),
 	}, nil
 }

--- a/cmd/erasure-object.go
+++ b/cmd/erasure-object.go
@@ -1475,37 +1475,15 @@ func (er erasureObjects) restoreTransitionedObject(ctx context.Context, bucket s
 		if err != nil {
 			return setRestoreHeaderFn(oi, err)
 		}
+		if pInfo.Size != partInfo.Size {
+			return setRestoreHeaderFn(oi, InvalidObjectState{Bucket: bucket, Object: object})
+		}
 		uploadedParts = append(uploadedParts, CompletePart{
 			PartNumber: pInfo.PartNumber,
 			ETag:       pInfo.ETag,
 		})
 	}
-
-	partsMatch := true
-	// validate parts created via multipart
-	if len(oi.Parts) == len(uploadedParts) {
-		for i, pi := range oi.Parts {
-			if uploadedParts[i].ETag != pi.ETag {
-				partsMatch = false
-				break
-			}
-		}
-	} else {
-		partsMatch = false
-	}
-
-	if !partsMatch {
-		return setRestoreHeaderFn(oi, InvalidObjectState{Bucket: bucket, Object: object})
-	}
-
 	_, err = er.CompleteMultipartUpload(ctx, bucket, object, uploadID, uploadedParts, ObjectOptions{
-		MTime:            oi.ModTime,
-		Versioned:        globalBucketVersioningSys.Enabled(bucket),
-		VersionSuspended: globalBucketVersioningSys.Suspended(bucket),
-	})
-	if err != nil {
-		uploadIDPath := er.getUploadIDDir(bucket, object, uploadID)
-		return setRestoreHeaderFn(oi, toObjectErr(err, minioMetaMultipartBucket, uploadIDPath))
-	}
-	return setRestoreHeaderFn(oi, nil)
+		MTime: oi.ModTime})
+	return setRestoreHeaderFn(oi, err)
 }

--- a/cmd/object-api-errors.go
+++ b/cmd/object-api-errors.go
@@ -495,7 +495,7 @@ func (e TransitionStorageClassNotFound) Error() string {
 type InvalidObjectState GenericError
 
 func (e InvalidObjectState) Error() string {
-	return "The operation is not valid for the current state of the object" + e.Bucket + "/" + e.Object
+	return "The operation is not valid for the current state of the object " + e.Bucket + "/" + e.Object + "(" + e.VersionID + ")"
 }
 
 /// Bucket related errors.

--- a/cmd/xl-storage.go
+++ b/cmd/xl-storage.go
@@ -1858,8 +1858,7 @@ func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath string, f
 	var srcDataPath string
 	var dstDataPath string
 	dataDir := retainSlash(fi.DataDir)
-	// no need to rename dataDir paths for objects that are in transitionComplete state.
-	if dataDir != "" && fi.TransitionStatus != lifecycle.TransitionComplete {
+	if dataDir != "" {
 		srcDataPath = retainSlash(pathJoin(srcVolumeDir, srcPath, dataDir))
 		// make sure to always use path.Join here, do not use pathJoin as
 		// it would additionally add `/` at the end and it comes in the

